src/us/kbase/auth2/lib/Authentication.java

```diff
@@ -1692,35 +1693,45 @@ public LoginToken loginProviderError(final String providerError) throws AuthStor
 	 * the URL will contain the identity provider authcode, and so the user should be redirected
 	 * to a new URL that does not contain said authcode as soon as possible.
 	 * 
+	 * @param token the temporary token returned by the {@link #loginStart(int, String, String)}
+	 * method.
 	 * @param provider the name of the identity provider that is servicing the login request.
 	 * @param authcode the authcode provided by the provider.
 	 * @param environment the environment in which the login request is occurring. Null for the
 	 * default environment.
 	 * @param tokenCtx the context under which the token will be created.
+	 * @param oauth2State the state value returned by the 3rd party identity provider.
 	 * @return either a login token or temporary token.
 	 * @throws MissingParameterException if the authcode is missing.
 	 * @throws IdentityRetrievalException if an error occurred when trying to retrieve identities
 	 * from the provider.
 	 * @throws AuthStorageException if an error occurred accessing the storage system.
 	 * @throws NoSuchIdentityProviderException if there is no provider by the given name.
-	 * @throws NoSuchEnvironmentException if no such environment is configured. 
+	 * @throws NoSuchEnvironmentException if no such environment is configured.
+	 * @throws InvalidTokenException if the token is invalid.
+	 * @throws AuthenticationException if the state doesn't match the expected state.
 	 */
-	public LoginToken login(
+	public LoginToken login( // enough args here to start considering a builder
+			final IncomingToken token,
 			final String provider,
 			final String authcode,
 			final String environment,
-			final TokenCreationContext tokenCtx)
+			final TokenCreationContext tokenCtx,
+			final String oauth2State)
 			throws MissingParameterException, IdentityRetrievalException,
-			AuthStorageException, NoSuchIdentityProviderException, NoSuchEnvironmentException {
-		nonNull(tokenCtx, "tokenCtx");
-		if (authcode == null || authcode.trim().isEmpty()) {
-			throw new MissingParameterException("authorization code");
-		}
+				AuthStorageException, NoSuchIdentityProviderException, NoSuchEnvironmentException,
+				InvalidTokenException, AuthenticationException {
+		requireNonNull(tokenCtx, "tokenCtx");
+		checkString(authcode, "authorization code");
```

Comment:
Looks familiar!

Reply:
Yeah, at some point I should package up all these little snippets into a jar, but they're so small it feels like leftpadding and I keep finding improvements over the old versions